src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/HubConnection.java

```diff
@@ -308,9 +309,10 @@ public Completable start() {
         negotiate.flatMapCompletable(url -> {
             logger.debug("Starting HubConnection.");
             if (transport == null) {
+                Single<String> tokenProvider = redirectAccessTokenProvider != null ? redirectAccessTokenProvider : accessTokenProvider;
```

Comment:
I noticed that we weren't nulling out the redirect url in stopConnection(), and saw we just flow the redirect url from the return value of startNegotiate() all the way to the call to transport.start(url).

I like this pattern over adding fields that need to be reset. I recommend changing startNegotiate() to return the NegotiateResponse returned from the last call to handleNegotiate().

startNegotiate() could continue to contain the transportEnum selection logic. Since getTransportEnum() exists, we need to store the current transportEnum in a field anyway. We *should* reset transportEnum to its original value in stopConnection().

The finalUrl logic currently in startNegotiate() should move here since this flatMapCompletable callback will have the entire last NegotiateResponse passed in, not just the finalUrl.

Then you can also move the access token logic that's currently in handleNegotiate() right here and use a local tokenProvider instead of the field. Speaking of that logic, I think we should remove the Authorization in stopConnection() if we added one.

Reply:
I don't disagree with these changes but are these refactorings something we want to include in the patch since this is something we want to backport? Just a thought.

Reply:
I'll just leave the refactor work our of the patch but I'll include it in master.

Reply:
>Then you can also move the access token logic that's currently in handleNegotiate() right here and use a local tokenProvider instead of the field.

I think we need to keep  a field for the accessToken because in the redirect scenario we'll get two negotiate responses. The accessToken won't be in the final negotiate response though.